### PR TITLE
[FileMoonIE] Add extractor for filemoon.sx

### LINKED ISSUE #31126: https://filemoon.sx/
## Checklist

- [ ✅] I'm reporting a new site support request
- [ ✅] I've verified that I'm running youtube-dl version **2021.12.17**
- [✅ ] I've checked that all provided URLs are alive and playable in a browser
- [ ✅] I've checked that none of provided URLs violate any copyrights
- [ ✅] I've searched the bugtracker for similar site support requests including closed ones

## Example URLs

- Single video: https://filemoon.sx/d/dzaq6p8yrvmm/Rick_Astley_-_Never_Gonna_Give_You_Up__Official_Music_Video___dQw4w9WgXcQ_.webm

## Description

Doesn't need an account I believe.

### PR DESCRIPTION
**Pull request by fonkap**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Adds support for filemoon.sx

Resolves #31126

### CODE REVIEW

File: youtube_dl/extractor/filemoon.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            'id': 'dw40rxrzruqz',
+            'title': 'dw40rxrzruqz',
+            'ext': 'mp4'
+        }
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        m = self._VALID_URL_RE.match(url)
+        title = m.group('title')
```
Comment (dirkf) [inline youtube_dl/extractor/filemoon.py:26]:
```suggestion
        video_id, title = re.match(self._VALID_URL, url).group('id', 'title')
```

File: youtube_dl/extractor/filemoon.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        if not title:
+            title = video_id
+
+        webpage = self._download_webpage(url, video_id)
+        matches = re.findall(r'(eval.*?)<\/script>', webpage, flags=re.DOTALL)
```
Comment (dirkf) [inline youtube_dl/extractor/filemoon.py:31]: Later you use `(?s)`. Better to go one way or the other:
```suggestion
        matches = re.findall(r'(?s)(eval.*?)<\/script>', webpage)
```

File: youtube_dl/extractor/filemoon.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        packed = matches[-1]
+        unpacked = utils.decode_packed_codes(packed)
+        jwplayer_sources = self._parse_json(
+            self._search_regex(
+                r"(?s)player\.setup\(\{sources:(.*?])", unpacked, 'jwplayer sources'),
```
Comment (dirkf) [inline youtube_dl/extractor/filemoon.py:36]: Relaxed:
```suggestion
                r'(?s)player\s*\.\s*setup\s*\(\s*\{\s*sources\s*:\s*(.*?])', unpacked, 'jwplayer sources'),
```

File: youtube_dl/extractor/filemoon.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        video_id = self._match_id(url)
+        m = self._VALID_URL_RE.match(url)
+        title = m.group('title')
+        if not title:
+            title = video_id
+
+        webpage = self._download_webpage(url, video_id)
+        matches = re.findall(r'(eval.*?)<\/script>', webpage, flags=re.DOTALL)
+        packed = matches[-1]
+        unpacked = utils.decode_packed_codes(packed)
```
Comment (dirkf) [inline youtube_dl/extractor/filemoon.py:33]:
```suggestion
        unpacked = decode_packed_codes(packed)
```

File: youtube_dl/extractor/filemoon.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+import re
+
+from .common import InfoExtractor
+from .. import utils
+from ..utils import js_to_json
```
Comment (dirkf) [inline youtube_dl/extractor/filemoon.py:8]:
```suggestion
from ..utils import (
    decode_packed_codes,
    js_to_json,
)
```

**Review by dirkf — CHANGES_REQUESTED** [review]
Thanks for work!

I've made some suggestions though probably can't all be committed without some more tweaks.

**Review by fonkap — COMMENTED** [review]
I applied all changes and they work fine.
Thanks @dirkf !

File: youtube_dl/extractor/filemoon.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+from .common import InfoExtractor
+from ..utils import (
+    decode_packed_codes,
+    js_to_json,
+)
+
+
+class FileMoonIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?filemoon\.sx/./(?P<id>\w+)(?:/(?P<title>.+))*'
```
Comment (dirkf) [inline youtube_dl/extractor/filemoon.py:14]:
```suggestion
    _VALID_URL = r'https?://(?:www\.)?filemoon\.sx/./(?P<id>\w+)'
```

File: youtube_dl/extractor/filemoon.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+            'id': 'dw40rxrzruqz',
+            'title': 'dw40rxrzruqz',
+            'ext': 'mp4'
+        }
+    }
+
+    def _real_extract(self, url):
+        video_id, title = re.match(self._VALID_URL, url).group('id', 'title')
+        if not title:
+            title = video_id
```
Comment (dirkf) [inline youtube_dl/extractor/filemoon.py:28]:
```suggestion
        video_id = self._match_id(url)
```

File: youtube_dl/extractor/filemoon.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        jwplayer_sources = self._parse_json(
+            self._search_regex(
+                r'(?s)player\s*\.\s*setup\s*\(\s*\{\s*sources\s*:\s*(.*?])', unpacked, 'jwplayer sources'),
+            video_id, transform_source=js_to_json)
+
+        formats = self._parse_jwplayer_formats(jwplayer_sources, video_id)
+
+        return {
+            'id': video_id,
+            'title': title,
```
Comment (dirkf) [inline youtube_dl/extractor/filemoon.py:43]:
```suggestion
            'title': self._generic_title(url) or video_id,
```

File: youtube_dl/extractor/filemoon.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+        }
+    }
+
+    def _real_extract(self, url):
+        video_id, title = re.match(self._VALID_URL, url).group('id', 'title')
+        if not title:
+            title = video_id
+
+        webpage = self._download_webpage(url, video_id)
+        matches = re.findall(r'(?s)(eval.*?)<\/script>', webpage)
```
Comment (dirkf) [inline youtube_dl/extractor/filemoon.py:31]:
```suggestion
        matches = re.findall(r'(?s)(eval.*?)</script>', webpage)
```

**Review by dirkf — CHANGES_REQUESTED** [review]
Just some simplification. I'm happy to apply these if you want and merge if CI tests are OK.

File: youtube_dl/extractor/filemoon.py (hunk 6 above)
Comment (fonkap) [reply youtube_dl/extractor/filemoon.py:14]: Not sure about this change, I know there are links with this form (filemoon/id/title) but I cannot find a very good example

Comment (dirkf) [reply youtube_dl/extractor/filemoon.py:14]: If there are links like that, the final part of the URL path will be picked up by `_generic_title()`, so replicating what you had, or I hope so.

**Review by fonkap — COMMENTED** [review]
All changes applied and working fine.
You were right! thanks

**Pull request merged**